File: client/typescript/test/dypthepenguin-test.ts
import { By, GamiumClient, KeyBy, NodeGamiumService, Player } from 'gamium';
import { test } from './functions';

(async () => {
  const gamiumService = new NodeGamiumService('127.0.0.1', 50061);
  await gamiumService.connect();
  const gamium = new GamiumClient(gamiumService);
  const ui = gamium.ui();
  const player = await gamium.player(By.path('/Dyp[1]/Dyp[1]/Root[1]'));
  const cameraLocator = By.path('/Cameras[1]/Main Camera[1]');

  await test('Pick First SnowBall', async () => {
    for (let index = 0; index < 3; index++) {
      const snowBall1Locator = By.path('/World[1]/Item Pickups[1]/SnowBallPickup[1]');
      const snowBall = await ui.find(snowBall1Locator);
      await snowBall.waitInteractable();
      await player.move(cameraLocator, snowBall1Locator);
      await gamium.sendKey(KeyBy.unityKeycode('E'));

      const snowBall2Locator = By.path('/World[1]/Item Pickups[1]/SnowBallPickup (1)[1]');
      const snowBall2 = await ui.find(snowBall2Locator);
      await snowBall2.waitInteractable();
      await player.move(cameraLocator, snowBall2Locator);
      await gamium.sendKey(KeyBy.unityKeycode('E'));
    }
  });

  await test('Pick Fish', async () => {
    await player.move(cameraLocator, { x: -1.74, y: 0.11, z: 0.07 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, By.path('/World[1]/Item Pickups[1]/FishPickup (1)[1]'));
    await gamium.sendKey(KeyBy.unityKeycode('E'));
    await gamium.sendKey(KeyBy.unityKeycode('Alpha2'));
  });

  await test('Kill Enemy1', async () => {
    await player.move(cameraLocator, { x: -19.66, y: -5.7, z: 28.51 });
    await player.move(cameraLocator, { x: -20.63, y: -5.7, z: 32.98 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: -24.56, y: -0.7, z: 41.63 });
    await gamium.sendKeys([KeyBy.unityKeycode('W'), KeyBy.unityKeycode('A')], { duratiomMs: 300 });
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
    await gamium.sleep(300);
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
  });

  await test('Move', async () => {
    await player.move(cameraLocator, { x: -13.83, y: -5.76, z: 21.76 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: -3.22, y: -2.37, z: 6.04 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 2.59, y: 0.11, z: 3.84 });
  });

  await test('Ride MovingPlatform1', async () => {
    await player.move(cameraLocator, { x: 4.88, y: 0.11, z: 0.84 });
    await player.move(cameraLocator, By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]'));
    await player.move(
      cameraLocator,
      By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]'),
      { epsilon: 1.6 },
    );
    await gamium.sendKey(KeyBy.unityKeycode('E'));
    for (let i = 0; i < Infinity; i++) {
      if (
        await player.isNear(
          By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/Mover Interactables[1]/Platform Mover Interactable (2)[1]/Crystal[1]'),
          { epsilon: 10 },
        )
      ) {
        break;
      }
      await gamium.sleep(100);
    }
    await gamium.sleep(2000);
  });

  await test('Kill Enemy2', async () => {
    await player.move(cameraLocator, { x: 20.88, y: -3.89, z: 24.6 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 26.59, y: -2.29, z: 33.18 });
    await gamium.sendKeys([KeyBy.unityKeycode('W'), KeyBy.unityKeycode('D')], { duratiomMs: 300 });
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
    await gamium.sleep(300);
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
  });

  await test('Ride MovingPlatform2', async () => {
    await player.move(cameraLocator, By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]'));
    await player.move(
      cameraLocator,
      By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]'),
      { epsilon: 1.6 },
    );
    await gamium.sendKey(KeyBy.unityKeycode('E'));
    for (let i = 0; i < Infinity; i++) {
      if (
        await player.isNear(
          By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/Mover Interactables[1]/Platform Mover Interactable (1)[1]/Crystal[1]'),
          { epsilon: 10 },
        )
      ) {
        break;
      }
      await gamium.sleep(100);
    }
    await gamium.sleep(2000);
  });

  await test('Jump Platform', async () => {
    await player.move(cameraLocator, { x: 9.15, y: 0.11, z: -0.99 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 19.14, y: -0.18, z: -0.19 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 29.34, y: -0.18, z: -0.35 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 39.33, y: -0.18, z: -0.67 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
  });

  await test('Kill Enemy3', async () => {
    for (let index = 0; index < 2; index++) {
      await player.move(cameraLocator, { x: 57.1, y: 0.11, z: 0.69 });
      await player.move(cameraLocator, { x: 47.74, y: 0.11, z: 1.34 });
      await gamium.sendKeys([KeyBy.unityKeycode('W'), KeyBy.unityKeycode('D')], { duratiomMs: 500 });
      await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
    }
    await player.move(cameraLocator, { x: 45.05, y: 0.11, z: 0.12 });
  });

  await test('Jump Platform back', async () => {
    await player.move(cameraLocator, { x: 44.11, y: -0.17, z: 0.22 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 33.91, y: -0.18, z: -0.78 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 23.87, y: -0.18, z: -0.56 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 14.1, y: -0.18, z: -0.91 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 6.84, y: 0.11, z: -0.45 });
  });

  await test('Go To Pick Key', async () => {
    await player.move(cameraLocator, { x: -1.81, y: -2.37, z: 15.22 });
    await player.move(cameraLocator, { x: 8.35, y: -8.8, z: 35.16 });
    await player.move(cameraLocator, { x: -3.17, y: -9, z: 45.8 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]'));
    await player.move(
      cameraLocator,
      By.path('/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]'),
      { epsilon: 1.6 },
    );
    await gamium.sendKey(KeyBy.unityKeycode('E'));
    for (let i = 0; i < Infinity; i++) {
      if (
        await player.isNear(
          By.path(
            '/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/Mover Interactables[1]/Platform Mover Interactable (2)[1]/Crystal[1]',
          ),
          { epsilon: 10 },
        )
      ) {
        break;
      }
      await gamium.sleep(100);
    }
    await gamium.sleep(2000);

    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, By.path('/World[1]/Item Pickups[1]/Special Pick Axe Pickup[1]'));
    await gamium.sendKey(KeyBy.unityKeycode('E'));
  });

  await test('Go to save Penguin', async () => {
    await player.move(cameraLocator, { x: 6.29, y: -8.8, z: 36.16 });
    await player.move(cameraLocator, { x: -0.47, y: -8.8, z: 35.29 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: -15.06, y: -5.7, z: 31.77 });
    await player.move(cameraLocator, { x: -10.58, y: -5.78, z: 23.14 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: -2.55, y: -2.37, z: 8.21 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
    await player.move(cameraLocator, { x: 2.02, y: 0.11, z: 1.58 });
    await player.move(cameraLocator, { x: -1.36, y: 0.11, z: 0.61 });
    await gamium.sendKey(KeyBy.unityKeycode('Space'));
  });

  await test('Save Penguin', async () => {
    await player.move(cameraLocator, By.path('/World[1]/World Environment[1]/Interactables[1]/Shackled Nice Penguin[1]/Nice Dyp[1]/penguin[1]'), {
      epsilon: 2.5,
    });
    await gamium.sendKey(KeyBy.unityKeycode('E'));
    await player.move(cameraLocator, { x: -12.61, y: 3.56, z: -1.43 }, { epsilon: 1.3 });
    await gamium.sendKeys([KeyBy.unityKeycode('W'), KeyBy.unityKeycode('A')], { duratiomMs: 100 });
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
    await gamium.sleep(1000);
    await player.move(cameraLocator, { x: -8.8, y: 3.56, z: 3.53 }, { epsilon: 1.3 });
    await gamium.sendKeys([KeyBy.unityKeycode('W'), KeyBy.unityKeycode('A')], { duratiomMs: 100 });
    await gamium.sendKey(KeyBy.unityKeycode('LeftControl'));
  });

  await test('Quit', async () => {
    await gamium.actions().appQuit().perform();
    process.exit(0);
  });
})().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: client/python/dypthepenguin-test.py
import uuid
from gamium import *
import asyncio
import sys


gamium = GamiumClient("127.0.0.1", 50061)
gamium.connect()
ui = gamium.ui()

player = gamium.player(By.path("/Dyp[1]/Dyp[1]/Root[1]"))
cameraLocator = By.path("/Cameras[1]/Main Camera[1]")


def pick_first_snowball():
    for index in range(3):
        snowBall1Locator = By.path("/World[1]/Item Pickups[1]/SnowBallPickup[1]")
        snowBall = ui.find(snowBall1Locator)
        snowBall.wait_interactable()
        player.move(cameraLocator, snowBall1Locator)
        gamium.send_key(KeyBy.unity_keycode("E"))

        snowBall2Locator = By.path("/World[1]/Item Pickups[1]/SnowBallPickup (1)[1]")
        snowBall2 = ui.find(snowBall2Locator)
        snowBall2.wait_interactable()
        player.move(cameraLocator, snowBall2Locator)
        gamium.send_key(KeyBy.unity_keycode("E"))


pick_first_snowball()


def pick_fish():
    player.move(cameraLocator, Vector3(-1.74, 0.11, 0.07))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, By.path("/World[1]/Item Pickups[1]/FishPickup (1)[1]"))
    gamium.send_key(KeyBy.unity_keycode("E"))
    gamium.send_key(KeyBy.unity_keycode("Alpha2"))


pick_fish()


def kill_enemy1():
    player.move(cameraLocator, Vector3(-19.66, -5.7, 28.51))
    player.move(cameraLocator, Vector3(-20.63, -5.7, 32.98))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(-24.56, -0.7, 41.63))
    gamium.send_keys([KeyBy.unity_keycode("W"), KeyBy.unity_keycode("A")], SendKeyOptions(300))
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))
    gamium.sleep(300)
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))


kill_enemy1()


def move():
    player.move(cameraLocator, Vector3(-13.83, -5.76, 21.76))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(-3.22, -2.37, 6.04))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(2.59, 0.11, 3.84))


move()


def ride_movingplatform1():
    player.move(cameraLocator, Vector3(4.88, 0.11, 0.84))
    player.move(
        cameraLocator, By.path("/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]")
    )
    player.move(
        cameraLocator,
        By.path(
            "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]"
        ),
        MovePlayerOptions(epsilon=1.6),
    )
    gamium.send_key(KeyBy.unity_keycode("E"))
    while True:
        if player.is_near(
            By.path(
                "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/Mover Interactables[1]/Platform Mover Interactable (2)[1]/Crystal[1]"
            ),
            10,
        ):
            break
        gamium.sleep(100)
    gamium.sleep(2000)


ride_movingplatform1()


def kill_enemy2():
    player.move(cameraLocator, Vector3(20.88, -3.89, 24.6))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(26.59, -2.29, 33.18))
    gamium.send_keys([KeyBy.unity_keycode("W"), KeyBy.unity_keycode("D")], SendKeyOptions(300))
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))
    gamium.sleep(300)
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))


kill_enemy2()


def ride_movingplatform2():
    player.move(
        cameraLocator, By.path("/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]")
    )
    player.move(
        cameraLocator,
        By.path(
            "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]"
        ),
        MovePlayerOptions(epsilon=1.6),
    )
    gamium.send_key(KeyBy.unity_keycode("E"))
    while True:
        if player.is_near(
            By.path(
                "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent[1]/Mover Interactables[1]/Platform Mover Interactable (1)[1]/Crystal[1]"
            ),
            10,
        ):
            break
        gamium.sleep(100)
    gamium.sleep(2000)


ride_movingplatform2()


def jump_platform():
    player.move(cameraLocator, Vector3(9.15, 0.11, -0.99))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(19.14, -0.18, -0.19))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(29.34, -0.18, -0.35))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(39.33, -0.18, -0.67))
    gamium.send_key(KeyBy.unity_keycode("Space"))


jump_platform()


def kill_enemy3():
    for index in range(2):
        player.move(cameraLocator, Vector3(57.1, 0.11, 0.69))
        player.move(cameraLocator, Vector3(47.74, 0.11, 1.34))
        gamium.send_keys([KeyBy.unity_keycode("W"), KeyBy.unity_keycode("D")], SendKeyOptions(500))
        gamium.send_key(KeyBy.unity_keycode("LeftControl"))
    player.move(cameraLocator, Vector3(45.05, 0.11, 0.12))


kill_enemy3()


def jump_platform_back():
    player.move(cameraLocator, Vector3(44.11, -0.17, 0.22))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(33.91, -0.18, -0.78))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(23.87, -0.18, -0.56))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(14.1, -0.18, -0.91))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(6.84, 0.11, -0.45))


jump_platform_back()


def go_to_pick_key():
    player.move(cameraLocator, Vector3(-1.81, -2.37, 15.22))
    player.move(cameraLocator, Vector3(8.35, -8.8, 35.16))
    player.move(cameraLocator, Vector3(-3.17, -9, 45.8))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(
        cameraLocator,
        By.path("/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/MovingPlatform[1]/M_Cylinder_Rounded[1]"),
    )
    player.move(
        cameraLocator,
        By.path(
            "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/MovingPlatform[1]/Platform Mover Interactable[1]/Crystal[1]"
        ),
        MovePlayerOptions(epsilon=1.6),
    )
    gamium.send_key(KeyBy.unity_keycode("E"))
    while True:
        if player.is_near(
            By.path(
                "/World[1]/World Environment[1]/Dynamic[1]/Moving Platforms[1]/MovingPlatform Parent (1)[1]/Mover Interactables[1]/Platform Mover Interactable (2)[1]/Crystal[1]",
            ),
            10,
        ):
            break
        gamium.sleep(100)
    gamium.sleep(2000)

    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, By.path("/World[1]/Item Pickups[1]/Special Pick Axe Pickup[1]"))
    gamium.send_key(KeyBy.unity_keycode("E"))


go_to_pick_key()


def go_to_save_penguin():
    player.move(cameraLocator, Vector3(6.29, -8.8, 36.16))
    player.move(cameraLocator, Vector3(-0.47, -8.8, 35.29))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(-15.06, -5.7, 31.77))
    player.move(cameraLocator, Vector3(-10.58, -5.78, 23.14))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(-2.55, -2.37, 8.21))
    gamium.send_key(KeyBy.unity_keycode("Space"))
    player.move(cameraLocator, Vector3(2.02, 0.11, 1.58))
    player.move(cameraLocator, Vector3(-1.36, 0.11, 0.61))
    gamium.send_key(KeyBy.unity_keycode("Space"))


go_to_save_penguin()


def save_penguin():
    player.move(
        cameraLocator,
        By.path("/World[1]/World Environment[1]/Interactables[1]/Shackled Nice Penguin[1]/Nice Dyp[1]/penguin[1]"),
        MovePlayerOptions(epsilon=2.5),
    )
    gamium.send_key(KeyBy.unity_keycode("E"))
    player.move(cameraLocator, Vector3(-12.61, 3.56, -1.43), MovePlayerOptions(epsilon=1.3))
    gamium.send_keys([KeyBy.unity_keycode("W"), KeyBy.unity_keycode("A")], SendKeyOptions(100))
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))
    gamium.sleep(1000)
    player.move(cameraLocator, Vector3(-8.8, 3.56, 3.53), MovePlayerOptions(epsilon=1.3))
    gamium.send_keys([KeyBy.unity_keycode("W"), KeyBy.unity_keycode("A")], SendKeyOptions(100))
    gamium.send_key(KeyBy.unity_keycode("LeftControl"))


save_penguin()


def quit():
    gamium.sleep(4000)
    gamium.actions().app_quit().perform()

    sys.exit(0)


quit()
